File: pegasusX/apps/warehouse-app-android/app/src/main/java/com/pegasusx/warehouse/ui/screens/drivers/DriversList.kt
package com.pegasusx.warehouse.ui.screens.drivers

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.grid.GridCells
import androidx.compose.foundation.lazy.grid.LazyVerticalGrid
import androidx.compose.foundation.lazy.grid.items
import androidx.compose.material3.*
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.pegasusx.warehouse.data.model.Driver
import com.pegasusx.warehouse.data.model.Vehicle
import com.pegasusx.warehouse.ui.components.WarehouseStatusChip
import com.pegasusx.warehouse.ui.theme.PegasusSpacing

private val DRIVER_UNAVAILABLE_REASON_LABELS = mapOf(
    "MAINTENANCE" to "Maintenance",
    "TRUCK_DAMAGED" to "Truck Damaged",
    "REGULATORY_HOLD" to "Regulatory Hold",
    "MANUAL_HOLD" to "Manual Hold",
)

@Composable
fun DriversList(
    drivers: List<Driver>,
    vehicles: List<Vehicle>,
    assigningDriverId: String?,
    onAssignClick: (Driver) -> Unit,
    modifier: Modifier = Modifier
) {
    LazyVerticalGrid(
        columns = GridCells.Adaptive(minSize = 340.dp),
        modifier = modifier.fillMaxSize(),
        contentPadding = PaddingValues(PegasusSpacing.lg),
        verticalArrangement = Arrangement.spacedBy(PegasusSpacing.md),
        horizontalArrangement = Arrangement.spacedBy(PegasusSpacing.md),
    ) {
        items(drivers, key = { it.driverId }) { driver ->
            ElevatedCard(modifier = Modifier.fillMaxWidth()) {
                Row(modifier = Modifier.padding(PegasusSpacing.lg), verticalAlignment = Alignment.CenterVertically) {
                    Column(modifier = Modifier.weight(1f)) {
                        Text(driver.name, style = MaterialTheme.typography.titleSmall)
                        Text(driver.phone, style = MaterialTheme.typography.bodySmall, color = MaterialTheme.colorScheme.onSurfaceVariant)
                        Text(
                            assignedVehicleLabel(driver, vehicles),
                            style = MaterialTheme.typography.bodySmall,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                        )
                        assignedVehicleReason(driver, vehicles)?.let { reason ->
                            Text(
                                reason,
                                style = MaterialTheme.typography.labelSmall,
                                color = MaterialTheme.colorScheme.tertiary,
                            )
                        }
                    }
                    Column(horizontalAlignment = Alignment.End, verticalArrangement = Arrangement.spacedBy(PegasusSpacing.sm)) {
                        WarehouseStatusChip(status = driver.truckStatus.ifBlank { "IDLE" })
                        OutlinedButton(
                            onClick = { onAssignClick(driver) },
                            enabled = assigningDriverId != driver.driverId,
                        ) {
                            if (assigningDriverId == driver.driverId) {
                                CircularProgressIndicator(modifier = Modifier.size(16.dp), strokeWidth = 2.dp)
                            } else {
                                Text(if (driver.vehicleId.isNullOrBlank()) "Assign" else "Reassign")
                            }
                        }
                    }
                }
            }
        }
    }
}

private fun assignedVehicleLabel(driver: Driver, vehicles: List<Vehicle>): String {
    val vehicleId = driver.vehicleId ?: return "Unassigned"
    val vehicle = vehicles.firstOrNull { it.vehicleId == vehicleId } ?: return "Assigned vehicle unavailable"
    return vehicleLabel(vehicle)
}

private fun assignedVehicleReason(driver: Driver, vehicles: List<Vehicle>): String? {
    val vehicleId = driver.vehicleId ?: return null
    if (!driver.vehicleIsActive) {
        return driver.vehicleUnavailableReason?.takeIf { it.isNotBlank() }
            ?.let { "Vehicle unavailable: ${vehicleUnavailableReasonLabel(it)}" }
            ?: "Vehicle unavailable"
    }
    val vehicle = vehicles.firstOrNull { it.vehicleId == vehicleId } ?: return null
    if (!vehicle.isActive) {
        return vehicle.unavailableReason?.takeIf { it.isNotBlank() }
            ?.let { "Vehicle unavailable: ${vehicleUnavailableReasonLabel(it)}" }
            ?: "Vehicle unavailable"
    }
    return null
}

fun vehicleLabel(vehicle: Vehicle): String {
    val title = if (vehicle.label.isBlank()) vehicle.licensePlate else vehicle.label
    return listOf(title, vehicle.vehicleClass).filter { it.isNotBlank() }.joinToString(" · ")
}

private fun vehicleUnavailableReasonLabel(reason: String): String {
    return DRIVER_UNAVAILABLE_REASON_LABELS[reason]
        ?: reason.lowercase().split('_').joinToString(" ") { token ->
            token.replaceFirstChar { ch -> ch.titlecase() }
        }
}

File: pegasusX/apps/warehouse-app-ios/WarehouseApp/Views/Drivers/DriversList.swift
import SwiftUI

struct DriversList: View {
    let drivers: [Driver]
    let vehicles: [Vehicle]
    let updatingDriverId: String?
    let onAssign: (String, String?) -> Void

    var body: some View {
        ResponsiveGridContentWrapper {
            ForEach(drivers) { driver in
                HStack {
                    VStack(alignment: .leading, spacing: LabTheme.spacingXS) {
                        Text(driver.name)
                            .font(.headline)
                        Text(driver.phone)
                            .font(.subheadline)
                            .foregroundStyle(.secondary)
                        Text(assignedVehicleLabel(for: driver))
                            .font(.caption)
                            .foregroundStyle(.secondary)
                        if let reason = assignedVehicleReason(for: driver), !reason.isEmpty {
                            Text(reason)
                                .font(.caption)
                                .foregroundStyle(.orange)
                        }
                    }
                    Spacer()
                    VStack(alignment: .trailing, spacing: LabTheme.spacingXS) {
                        Text(driver.truckStatus.isEmpty ? "IDLE" : driver.truckStatus)
                            .font(.caption.bold())
                            .padding(.horizontal, LabTheme.spacingSM)
                            .padding(.vertical, LabTheme.spacingXS)
                            .background(.quaternary, in: Capsule())
                        Menu {
                            Button("Unassign") {
                                onAssign(driver.driverId, nil)
                            }
                            ForEach(assignableVehicles(for: driver)) { vehicle in
                                Button(vehicleLabel(for: vehicle)) {
                                    onAssign(driver.driverId, vehicle.vehicleId)
                                }
                            }
                        } label: {
                            if updatingDriverId == driver.driverId {
                                ProgressView()
                                    .controlSize(.small)
                            } else {
                                Label(driver.vehicleId == nil ? "Assign" : "Reassign", systemImage: "truck.box")
                                    .font(.caption)
                            }
                        }
                        .disabled(updatingDriverId == driver.driverId)
                    }
                }
            }
        }
    }

    private func assignedVehicleLabel(for driver: Driver) -> String {
        guard let vehicleId = driver.vehicleId,
              let vehicle = vehicles.first(where: { $0.vehicleId == vehicleId }) else {
            return "Unassigned"
        }
        return vehicleLabel(for: vehicle)
    }

    private func assignedVehicleReason(for driver: Driver) -> String? {
        guard driver.vehicleId != nil else {
            return nil
        }

        if driver.vehicleIsActive == false {
            if let reason = driver.vehicleUnavailableReason, !reason.isEmpty {
                return "Vehicle unavailable: \(vehicleUnavailableReasonLabel(reason))"
            }
            return "Vehicle unavailable"
        }

        guard let vehicleId = driver.vehicleId,
              let vehicle = vehicles.first(where: { $0.vehicleId == vehicleId }),
              !vehicle.isActive else {
            return nil
        }

        if let reason = vehicle.unavailableReason, !reason.isEmpty {
            return "Vehicle unavailable: \(vehicleUnavailableReasonLabel(reason))"
        }
        return "Vehicle unavailable"
    }

    private func assignableVehicles(for driver: Driver) -> [Vehicle] {
        vehicles.filter { $0.isActive || $0.vehicleId == driver.vehicleId }
    }

    private func vehicleLabel(for vehicle: Vehicle) -> String {
        let title = vehicle.label.isEmpty ? vehicle.licensePlate : vehicle.label
        return [title, vehicle.vehicleClass].filter { !$0.isEmpty }.joined(separator: " · ")
    }
}
